File: advent_of_code/2023/day02/Day2.py
import os

class Day2:
    def __init__(self):
        self.input_content = None

    def solve_part1(self):
        sum = 0
        colours = ["red", "green", "blue"]

        for line in self.input_content.splitlines():
            count = [0, 0, 0]
            id = int(line.split(":")[0].split(" ")[1])

            line = line.split(":")[1]
            for games in line.split(";"):
                entries = games.split(",")
                for entry in entries:
                    entry = entry.strip()
                    curr = int(entry.split(" ")[0])
                    colour = entry.split(" ")[1]

                    if colour in colours:
                        index = colours.index(colour)
                        count[index] = max(count[index], curr)
            
            if count[0] <= 12 and count[1] <= 13 and count[2] <= 14:
                sum += id

        return sum

    def solve_part2(self):
        sets = []
        colours = ["red", "green", "blue"]
        
        for line in self.input_content.splitlines():
            count = [0, 0, 0]
            id = int(line.split(":")[0].split(" ")[1])

            line = line.split(":")[1]
            for games in line.split(";"):
                entries = games.split(",")
                for entry in entries:
                    entry = entry.strip()
                    curr = int(entry.split(" ")[0])
                    colour = entry.split(" ")[1]

                    if colour in colours:
                        index = colours.index(colour)
                        count[index] = max(count[index], curr)
            
            sets.append(count)
        
        sum = 0
        for i in range(len(sets)):
            curr = sets[i]
            power = sets[i][0] * sets[i][1] * sets[i][2]
            sum += power
            
        return sum

    def loadInputFiles(self):
        inputPath = os.path.join(os.getcwd(), "advent_of_code", "2023", "day2", "input.txt")
        with open(inputPath, "r") as f:
            self.input_content = f.read()

# Usage example:
solver = Day2()
solver.loadInputFiles()
part1_result = solver.solve_part1()
part2_result = solver.solve_part2()

print("Solution to Part 1:", part1_result)
print("Solution to Part 2:", part2_result)

File: advent_of_code/2023/day02/java/Day2.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

public class Day2 {

    public static void main(String[] args) throws IOException {
        
        List<String> input = Files.readAllLines(Path.of("input.txt"));
        int sum = 0;
        int power = 0;

        for(String game : input) {
            int id = Integer.parseInt(game.split(":")[0].split(" ")[1]);
            int[] counts = new int[]{0, 0, 0};

            String[] games = game.split(":")[1].split(";");

            for(String entry : games) {
                String[] sets = entry.trim().split(",");
                for(String set : sets){
                    String[] balls = set.trim().split(" ");

                    int number = Integer.parseInt(balls[0]);
                    String colour = balls[1];

                    if(colour.equals("red")) {
                        counts[0] = Math.max(counts[0], number);
                    } else if(colour.equals("green")) {
                        counts[1] = Math.max(counts[1], number);
                    } else if(colour.equals("blue")) {
                        counts[2] = Math.max(counts[2], number);
                    }
                }
            }
            
            if(counts[0] <= 12 && counts[1] <= 13 && counts[2] <= 14) {
                sum += id;
            }

            power += counts[0] * counts[1] * counts[2];
        }

        System.out.println("part 1: " + sum);
        System.out.println("part 2: " + power);
    }

}